ingestion-beam/src/main/java/com/mozilla/telemetry/io/Write.java

```diff
@@ -148,6 +173,125 @@ public FileOutput(ValueProvider<String> outputPrefix, OutputFileFormat format,
+   */
+  public static class AvroOutput extends Write {
+
+    private final ValueProvider<String> outputPrefix;
+    private final Duration windowDuration;
+    private final ValueProvider<Integer> numShards;
+    private final Compression compression;
+    private final InputType inputType;
+    private final AvroSchemaStore schemaStore;
+
+    /** Public constructor. */
+    public AvroOutput(ValueProvider<String> outputPrefix, Duration windowDuration,
+        ValueProvider<Integer> numShards, Compression compression, InputType inputType,
+        ValueProvider<String> schemasLocation) {
+      this.outputPrefix = outputPrefix;
+      this.windowDuration = windowDuration;
+      this.numShards = numShards;
+      this.compression = compression;
+      this.inputType = inputType;
+      this.schemaStore = AvroSchemaStore.of(schemasLocation);
+    }
+
+    /** A functor that constructs a GenericRecord from a PubsubMessage. */
+    private class RecordFormatter implements AvroIO.RecordFormatter<PubsubMessage> {
+
+      public GenericRecord formatRecord(PubsubMessage element, Schema schema) {
+        String message;
+        GenericRecord result;
+        try {
+          message = Json.asString(element);
+          DatumReader<GenericRecord> reader = new GenericDatumReader<GenericRecord>(schema);
+          Decoder decoder = DecoderFactory.get().jsonDecoder(schema, message);
+          result = reader.read(null, decoder);
+        } catch (IOException e) {
+          // We rely on PubsubMessage construction to fail on invalid data, so we should
+          // never
+          // see a non-encodable PubsubMessage here and we let the exception bubble up if
+          // it happens.
+          throw new UncheckedIOException(e);
+        }
+        return result;
+      }
+    }
+
+    class PubsubMessageDynamicAvroDestinations
+        extends DynamicAvroDestinations<PubsubMessage, String, GenericRecord> {
+
+      final PCollectionView<AvroSchemaStore> schemaStore;
+      final RecordFormatter formatter;
+      ValueProvider<DynamicPathTemplate> pathTemplate = NestedValueProvider.of(outputPrefix,
+          DynamicPathTemplate::new);
+      ValueProvider<String> staticPrefix = NestedValueProvider.of(pathTemplate,
+          value -> value.staticPrefix);
+
+      public PubsubMessageDynamicAvroDestinations(PCollectionView<AvroSchemaStore> schemaStore) {
+        this.schemaStore = schemaStore;
+        this.formatter = new RecordFormatter();
+      }
+
+      public GenericRecord formatRecord(PubsubMessage message) {
+        // TODO: unhandled exception
+        Schema schema = sideInput(schemaStore).getSchema(message.getAttributeMap());
+        return formatter.formatRecord(message, schema);
+      }
+
+      public Schema getSchema(String schemaPath) {
+        // TODO: unhandled exception
+        return sideInput(schemaStore).getSchema(schemaPath);
```

Comment:
I haven't figured out what to do with unhandled `SchemaNotFoundException`s, it doesn't seem like there's a way to bubble up the exceptions into the pipeline.